### DISCUSSION TITLE
请问下云大,lua5.4.4 ltable.c文件中源码这块为什么进入这个地方的时候要cast_int(ui)一下呢，看注释说这样更快,实在没理解

### DISCUSSION THREAD

**Discussion by frog-game** [cloudwu/skynet#1697] [General]
```c
/*
** Hash for integers. To allow a good hash, use the remainder operator
** ('%'). If integer fits as a non-negative int, compute an int
** remainder, which is faster. Otherwise, use an unsigned-integer
** remainder, which uses all bits and ensures a non-negative result.
*/

/// @brief 返回对应整数值对散列表大小取余的key对应的节点 
/// @param t 
/// @param i key
/// @return 
static Node *hashint (const Table *t, lua_Integer i) {
  lua_Unsigned ui = l_castS2U(i);//如果要对负数计算哈希值的话，先转成正数方便计算
  if (ui <= (unsigned int)INT_MAX)//没有超过无符号int最大值 (unsigned int)INT_MAX = 2147483647
    return hashmod(t, cast_int(ui));  //--------------疑问点----------------------------       
  else
    return hashmod(t, ui);
}
```

**Comment by cloudwu** [comment]
你投错地方了。你应该发到 lua mailling list .

**Reply by frog-game** [reply]
好的~

**Comment by frog-game** [comment]
https://listmaster.pepperfish.net/hyperkitty/list/lua-l@lists.lua.org/thread/APOVIQG3GZZR7F7ZLN4FI2BPHYM6VARP/
发个我在lua mailling list 上热心网友bil til和lua作者的Roberto回答过程。。


copy that, thank you~~

Roberto Ierusalimschy <roberto(a)inf.puc-rio.br&gt; 于2023年1月21日周六 01:46写道：
>
> > I found this explanation on this webpage
> >
[https://answerbun.com/stack-overflow/why-is-__int128_t-faster-than-long-l...](https://answerbun.com/stack-overflow/why-is-__int128_t-faster-than-long-long-on-x86-64-gcc/)
> >
> > idiv r64 is 56 uops for the front-end, with latency from 41 to 95
> > cycles (from divisor to quotient, which is the relevant case here I
> > think).
> > div r64 is 33 uops for the front-end, with latency from 35 to 87
> > cycles. (for that same latency path).
> >
> > Probably DIV is faster than iDIV,
> > the DIV latency is shorter
>
> Precisely. As the comment there explains: "If integer fits as a
> non-negative int, compute an int remainder, which is faster. Otherwise,
> use an unsigned-integer remainder, which uses all bits and ensures a
> non-negative result."
>
> -- Roberto

结个贴
感谢他们的辛苦指点~
